File: repertoire/concepts/data-integration/providers/enricher_plugin/ner_extract.ts
// Clef Data Integration Kit - Named Entity Recognition enricher provider
// Tokenizes text, applies NER model/rules (pattern-based for known entity types), returns entity spans.

export const PROVIDER_ID = 'ner_extract';
export const PLUGIN_TYPE = 'enricher_plugin';

export interface ContentItem {
  id: string;
  content: string;
  contentType: string;
  metadata?: Record<string, unknown>;
}

export interface EnricherConfig {
  model?: string;
  apiKey?: string;
  threshold?: number;
  options?: Record<string, unknown>;
}

export interface EnrichmentResult {
  fields: Record<string, unknown>;
  confidence: number;
  metadata?: Record<string, unknown>;
}

export interface SchemaRef {
  name: string;
  fields?: string[];
}

export interface CostEstimate {
  tokens?: number;
  apiCalls?: number;
  durationMs?: number;
}

type EntityType = 'PERSON' | 'ORG' | 'LOC' | 'EVENT' | 'DATE' | 'MONEY' | 'EMAIL' | 'URL' | 'PHONE';

interface Entity {
  text: string;
  type: EntityType;
  start: number;
  end: number;
  confidence: number;
}

// Pattern-based NER rules for common entity types
const ENTITY_PATTERNS: { type: EntityType; pattern: RegExp; confidence: number }[] = [
  // EMAIL: standard email pattern
  { type: 'EMAIL', pattern: /\b[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}\b/g, confidence: 0.98 },
  // URL: http/https URLs
  { type: 'URL', pattern: /https?:\/\/[^\s<>\"']+/g, confidence: 0.97 },
  // PHONE: various phone formats
  { type: 'PHONE', pattern: /(?:\+\d{1,3}[-.\s]?)?\(?\d{2,4}\)?[-.\s]?\d{3,4}[-.\s]?\d{3,4}\b/g, confidence: 0.85 },
  // DATE: common date patterns
  { type: 'DATE', pattern: /\b(?:(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{1,2},?\s+\d{4}|\d{1,2}\/\d{1,2}\/\d{2,4}|\d{4}-\d{2}-\d{2})\b/gi, confidence: 0.92 },
  // MONEY: currency amounts
  { type: 'MONEY', pattern: /(?:\$|EUR|GBP|USD|JPY)\s?\d{1,3}(?:,\d{3})*(?:\.\d{1,2})?(?:\s?(?:million|billion|thousand|M|B|K))?\b/gi, confidence: 0.9 },
];

// Title-cased word sequences indicate PERSON, ORG, LOC candidates
const TITLE_CASE_SEQUENCE = /\b([A-Z][a-z]+(?:\s+(?:of|the|and|de|van|von|al|el)\s+)?(?:[A-Z][a-z]+\s*){0,4})\b/g;

// Context clue words for disambiguation
const PERSON_INDICATORS = new Set([
  'mr', 'mrs', 'ms', 'dr', 'prof', 'sir', 'president', 'ceo', 'director',
  'said', 'told', 'according', 'born', 'died', 'married', 'senator', 'minister',
]);
const ORG_INDICATORS = new Set([
  'inc', 'corp', 'ltd', 'llc', 'company', 'organization', 'foundation',
  'university', 'institute', 'bank', 'group', 'association', 'commission',
]);
const LOC_INDICATORS = new Set([
  'city', 'state', 'country', 'river', 'mountain', 'island', 'street',
  'avenue', 'boulevard', 'county', 'province', 'district', 'republic',
]);
const EVENT_INDICATORS = new Set([
  'conference', 'summit', 'festival', 'championship', 'olympics', 'election',
  'ceremony', 'tournament', 'expo', 'convention', 'meeting', 'war', 'battle',
]);

function getContextWords(text: string, start: number, end: number, windowSize: number): string[] {
  const before = text.slice(Math.max(0, start - windowSize), start).toLowerCase();
  const after = text.slice(end, Math.min(text.length, end + windowSize)).toLowerCase();
  return (before + ' ' + after).split(/\s+/).filter(Boolean);
}

function classifyTitleCaseEntity(
  text: string,
  fullContent: string,
  start: number,
  end: number,
): { type: EntityType; confidence: number } | null {
  const contextWords = getContextWords(fullContent, start, end, 100);
  const entityWords = text.toLowerCase().split(/\s+/);

  // Check organization indicators (suffix-based are high confidence)
  const orgSuffixes = ['inc', 'corp', 'ltd', 'llc', 'co'];
  if (orgSuffixes.some((s) => entityWords[entityWords.length - 1] === s)) {
    return { type: 'ORG', confidence: 0.92 };
  }
  if (contextWords.some((w) => ORG_INDICATORS.has(w)) || entityWords.some((w) => ORG_INDICATORS.has(w))) {
    return { type: 'ORG', confidence: 0.78 };
  }

  // Check person indicators
  if (contextWords.some((w) => PERSON_INDICATORS.has(w))) {
    return { type: 'PERSON', confidence: 0.82 };
  }
  // 2-3 word title case with no org/loc clue often indicates person
  if (entityWords.length >= 2 && entityWords.length <= 3) {
    const allCapitalized = text.split(/\s+/).every((w) => /^[A-Z]/.test(w));
    if (allCapitalized) return { type: 'PERSON', confidence: 0.65 };
  }

  // Check location indicators
  if (contextWords.some((w) => LOC_INDICATORS.has(w)) || entityWords.some((w) => LOC_INDICATORS.has(w))) {
    return { type: 'LOC', confidence: 0.75 };
  }

  // Check event indicators
  if (contextWords.some((w) => EVENT_INDICATORS.has(w)) || entityWords.some((w) => EVENT_INDICATORS.has(w))) {
    return { type: 'EVENT', confidence: 0.7 };
  }

  // If no strong signal, skip low-confidence guesses
  return null;
}

function deduplicateEntities(entities: Entity[]): Entity[] {
  // Remove overlapping entities, preferring higher confidence
  entities.sort((a, b) => b.confidence - a.confidence);
  const result: Entity[] = [];
  for (const entity of entities) {
    const overlaps = result.some(
      (e) => entity.start < e.end && entity.end > e.start,
    );
    if (!overlaps) {
      result.push(entity);
    }
  }
  return result.sort((a, b) => a.start - b.start);
}

export class NerExtractEnricherProvider {
  async enrich(item: ContentItem, config: EnricherConfig): Promise<EnrichmentResult> {
    const entityTypes = (config.options?.entityTypes as EntityType[]) ?? [
      'PERSON', 'ORG', 'LOC', 'EVENT', 'DATE', 'MONEY', 'EMAIL', 'URL', 'PHONE',
    ];
    const threshold = config.threshold ?? 0.5;
    const text = item.content;

    const entities: Entity[] = [];

    // Apply regex-based patterns for structured entity types
    for (const { type, pattern, confidence } of ENTITY_PATTERNS) {
      if (!entityTypes.includes(type)) continue;
      pattern.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = pattern.exec(text)) !== null) {
        entities.push({
          text: match[0],
          type,
          start: match.index,
          end: match.index + match[0].length,
          confidence,
        });
      }
    }

    // Apply title-case sequence matching for PERSON, ORG, LOC, EVENT
    const titleTypes: EntityType[] = ['PERSON', 'ORG', 'LOC', 'EVENT'];
    if (titleTypes.some((t) => entityTypes.includes(t))) {
      TITLE_CASE_SEQUENCE.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = TITLE_CASE_SEQUENCE.exec(text)) !== null) {
        const candidateText = match[1].trim();
        // Skip single common words that happen to be capitalized at sentence start
        if (candidateText.split(/\s+/).length < 2 && match.index > 0 && text[match.index - 1] !== '.') {
          continue;
        }

        const classification = classifyTitleCaseEntity(
          candidateText, text, match.index, match.index + candidateText.length,
        );
        if (classification && entityTypes.includes(classification.type)) {
          entities.push({
            text: candidateText,
            type: classification.type,
            start: match.index,
            end: match.index + candidateText.length,
            confidence: classification.confidence,
          });
        }
      }
    }

    // Filter by threshold and deduplicate
    const filtered = deduplicateEntities(
      entities.filter((e) => e.confidence >= threshold),
    );

    const avgConfidence = filtered.length > 0
      ? filtered.reduce((sum, e) => sum + e.confidence, 0) / filtered.length
      : 0;

    return {
      fields: {
        entities: filtered,
        entity_count: filtered.length,
        entity_type_counts: entityTypes.reduce((acc, type) => {
          acc[type] = filtered.filter((e) => e.type === type).length;
          return acc;
        }, {} as Record<string, number>),
      },
      confidence: avgConfidence,
      metadata: {
        provider: PROVIDER_ID,
        entityTypes,
        threshold,
        mode: 'pattern_based',
      },
    };
  }

  appliesTo(schema: SchemaRef): boolean {
    const textSchemas = ['text', 'article', 'document', 'content', 'post', 'message', 'note'];
    return textSchemas.some((s) => schema.name.toLowerCase().includes(s));
  }

  costEstimate(item: ContentItem): CostEstimate {
    const charCount = item.content.length;
    // Pattern-based NER: ~0.1ms per 100 chars
    const durationMs = Math.max(10, Math.ceil(charCount / 1000));
    return { durationMs, apiCalls: 0 };
  }
}

export default NerExtractEnricherProvider;

File: repertoire/concepts/data-integration/providers/enricher_plugin/ner_extract.rs
// Clef Data Integration Kit - Named Entity Recognition enricher provider
// Tokenizes text, applies NER rules (pattern-based for known entity types), returns entity spans.

use std::collections::HashMap;

pub const PROVIDER_ID: &str = "ner_extract";
pub const PLUGIN_TYPE: &str = "enricher_plugin";

#[derive(Debug, Clone)]
pub struct ContentItem {
    pub id: String,
    pub content: String,
    pub content_type: String,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct EnricherConfig {
    pub model: Option<String>,
    pub api_key: Option<String>,
    pub threshold: Option<f64>,
    pub options: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct EnrichmentResult {
    pub fields: HashMap<String, serde_json::Value>,
    pub confidence: f64,
    pub metadata: Option<HashMap<String, serde_json::Value>>,
}

#[derive(Debug, Clone)]
pub struct SchemaRef {
    pub name: String,
    pub fields: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub struct CostEstimate {
    pub tokens: Option<u64>,
    pub api_calls: Option<u64>,
    pub duration_ms: Option<u64>,
}

#[derive(Debug, Clone, PartialEq, serde::Serialize)]
pub enum EntityType {
    PERSON,
    ORG,
    LOC,
    EVENT,
    DATE,
    MONEY,
    EMAIL,
    URL,
    PHONE,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct Entity {
    pub text: String,
    pub entity_type: EntityType,
    pub start: usize,
    pub end: usize,
    pub confidence: f64,
}

#[derive(Debug)]
pub enum EnricherError {
    ParseError(String),
}

const PERSON_INDICATORS: &[&str] = &[
    "mr", "mrs", "ms", "dr", "prof", "sir", "president", "ceo", "director",
    "said", "told", "according", "born", "died", "married", "senator",
];
const ORG_INDICATORS: &[&str] = &[
    "inc", "corp", "ltd", "llc", "company", "organization", "foundation",
    "university", "institute", "bank", "group", "association", "commission",
];
const LOC_INDICATORS: &[&str] = &[
    "city", "state", "country", "river", "mountain", "island", "street",
    "avenue", "boulevard", "county", "province", "district", "republic",
];
const EVENT_INDICATORS: &[&str] = &[
    "conference", "summit", "festival", "championship", "olympics", "election",
    "ceremony", "tournament", "expo", "convention", "meeting", "war", "battle",
];

fn find_email_entities(text: &str) -> Vec<Entity> {
    let mut entities = Vec::new();
    let bytes = text.as_bytes();
    let len = bytes.len();
    let mut i = 0;

    while i < len {
        if bytes[i] == b'@' && i > 0 {
            // Scan backwards for local part
            let mut local_start = i;
            while local_start > 0 {
                let c = bytes[local_start - 1];
                if c.is_ascii_alphanumeric() || b"._%+-".contains(&c) {
                    local_start -= 1;
                } else {
                    break;
                }
            }
            // Scan forwards for domain
            let mut domain_end = i + 1;
            let mut has_dot = false;
            while domain_end < len {
                let c = bytes[domain_end];
                if c.is_ascii_alphanumeric() || c == b'.' || c == b'-' {
                    if c == b'.' { has_dot = true; }
                    domain_end += 1;
                } else {
                    break;
                }
            }
            if has_dot && domain_end > i + 1 && local_start < i {
                entities.push(Entity {
                    text: text[local_start..domain_end].to_string(),
                    entity_type: EntityType::EMAIL,
                    start: local_start,
                    end: domain_end,
                    confidence: 0.98,
                });
            }
        }
        i += 1;
    }
    entities
}

fn find_url_entities(text: &str) -> Vec<Entity> {
    let mut entities = Vec::new();
    let mut search_from = 0;
    while let Some(pos) = text[search_from..].find("http") {
        let abs_pos = search_from + pos;
        let remaining = &text[abs_pos..];
        if remaining.starts_with("http://") || remaining.starts_with("https://") {
            let end = remaining.find(|c: char| c.is_whitespace() || "<>\"'".contains(c))
                .unwrap_or(remaining.len());
            entities.push(Entity {
                text: remaining[..end].to_string(),
                entity_type: EntityType::URL,
                start: abs_pos,
                end: abs_pos + end,
                confidence: 0.97,
            });
            search_from = abs_pos + end;
        } else {
            search_from = abs_pos + 4;
        }
    }
    entities
}

fn find_date_entities(text: &str) -> Vec<Entity> {
    let mut entities = Vec::new();
    let months = [
        "january", "february", "march", "april", "may", "june",
        "july", "august", "september", "october", "november", "december",
    ];
    let lower = text.to_lowercase();

    for month in &months {
        let mut pos = 0;
        while let Some(idx) = lower[pos..].find(month) {
            let abs_idx = pos + idx;
            // Look for "Month DD, YYYY" or "Month DD YYYY"
            let after_month = &text[abs_idx + month.len()..];
            let after_trimmed = after_month.trim_start();
            let digits_and_more: String = after_trimmed.chars()
                .take_while(|c| c.is_ascii_digit() || *c == ',' || *c == ' ')
                .collect();
            let numbers: Vec<&str> = digits_and_more.split(|c: char| !c.is_ascii_digit())
                .filter(|s| !s.is_empty())
                .collect();

            if numbers.len() >= 2 {
                let day: u32 = numbers[0].parse().unwrap_or(0);
                let year: u32 = numbers[1].parse().unwrap_or(0);
                if day >= 1 && day <= 31 && year >= 1900 && year <= 2100 {
                    let end_offset = abs_idx + month.len() + digits_and_more.len() + (after_month.len() - after_trimmed.len());
                    entities.push(Entity {
                        text: text[abs_idx..end_offset].trim().to_string(),
                        entity_type: EntityType::DATE,
                        start: abs_idx,
                        end: end_offset,
                        confidence: 0.92,
                    });
                }
            }
            pos = abs_idx + month.len();
        }
    }

    // ISO date pattern: YYYY-MM-DD
    let chars: Vec<char> = text.chars().collect();
    for i in 0..chars.len().saturating_sub(9) {
        if chars[i].is_ascii_digit() && chars[i+4] == '-' && chars[i+7] == '-' {
            let candidate: String = chars[i..i+10].iter().collect();
            if candidate.len() == 10 {
                let parts: Vec<&str> = candidate.split('-').collect();
                if parts.len() == 3 {
                    let y: u32 = parts[0].parse().unwrap_or(0);
                    let m: u32 = parts[1].parse().unwrap_or(0);
                    let d: u32 = parts[2].parse().unwrap_or(0);
                    if y >= 1900 && y <= 2100 && m >= 1 && m <= 12 && d >= 1 && d <= 31 {
                        entities.push(Entity {
                            text: candidate,
                            entity_type: EntityType::DATE,
                            start: i,
                            end: i + 10,
                            confidence: 0.95,
                        });
                    }
                }
            }
        }
    }
    entities
}

fn find_title_case_entities(text: &str, threshold: f64) -> Vec<Entity> {
    let mut entities = Vec::new();
    let words: Vec<(usize, &str)> = text.split_whitespace()
        .scan(0usize, |pos, word| {
            let start = text[*pos..].find(word).map(|i| *pos + i).unwrap_or(*pos);
            *pos = start + word.len();
            Some((start, word))
        })
        .collect();

    let mut i = 0;
    while i < words.len() {
        let (start, word) = words[i];
        if word.len() > 1 && word.chars().next().map_or(false, |c| c.is_uppercase())
            && word.chars().skip(1).any(|c| c.is_lowercase()) {
            // Collect consecutive title-case words (allow connectors)
            let connectors = ["of", "the", "and", "de", "van", "von", "al", "el"];
            let mut j = i + 1;
            while j < words.len() {
                let next_word = words[j].1;
                if connectors.contains(&next_word) && j + 1 < words.len() {
                    j += 1;
                    continue;
                }
                if next_word.len() > 1 && next_word.chars().next().map_or(false, |c| c.is_uppercase())
                    && next_word.chars().skip(1).any(|c| c.is_lowercase()) {
                    j += 1;
                } else {
                    break;
                }
            }

            if j > i + 1 || (i > 0 && !text[..start].ends_with(". ")) {
                let end = words[j - 1].0 + words[j - 1].1.len();
                let entity_text = text[start..end].trim().to_string();
                let context = get_context_words(text, start, end, 100);
                let entity_words: Vec<&str> = entity_text.to_lowercase().split_whitespace().collect();

                // Classify based on context
                let last_word = entity_words.last().map(|s| *s).unwrap_or("");
                let org_suffixes = ["inc", "corp", "ltd", "llc", "co"];

                let classification = if org_suffixes.contains(&last_word) {
                    Some((EntityType::ORG, 0.92))
                } else if context.iter().any(|w| ORG_INDICATORS.contains(&w.as_str())) {
                    Some((EntityType::ORG, 0.78))
                } else if context.iter().any(|w| PERSON_INDICATORS.contains(&w.as_str())) {
                    Some((EntityType::PERSON, 0.82))
                } else if entity_words.len() >= 2 && entity_words.len() <= 3 {
                    Some((EntityType::PERSON, 0.65))
                } else if context.iter().any(|w| LOC_INDICATORS.contains(&w.as_str())) {
                    Some((EntityType::LOC, 0.75))
                } else if context.iter().any(|w| EVENT_INDICATORS.contains(&w.as_str())) {
                    Some((EntityType::EVENT, 0.70))
                } else {
                    None
                };

                if let Some((etype, conf)) = classification {
                    if conf >= threshold {
                        entities.push(Entity {
                            text: entity_text,
                            entity_type: etype,
                            start, end,
                            confidence: conf,
                        });
                    }
                }
            }
            i = j;
        } else {
            i += 1;
        }
    }
    entities
}

fn get_context_words(text: &str, start: usize, end: usize, window: usize) -> Vec<String> {
    let before_start = start.saturating_sub(window);
    let after_end = (end + window).min(text.len());
    let context = format!("{} {}", &text[before_start..start], &text[end..after_end]);
    context.to_lowercase().split_whitespace().map(String::from).collect()
}

fn deduplicate_entities(mut entities: Vec<Entity>) -> Vec<Entity> {
    entities.sort_by(|a, b| b.confidence.partial_cmp(&a.confidence).unwrap_or(std::cmp::Ordering::Equal));
    let mut result = Vec::new();
    for entity in entities {
        let overlaps = result.iter().any(|e: &Entity| entity.start < e.end && entity.end > e.start);
        if !overlaps {
            result.push(entity);
        }
    }
    result.sort_by_key(|e| e.start);
    result
}

pub struct NerExtractEnricherProvider;

impl NerExtractEnricherProvider {
    pub fn new() -> Self {
        Self
    }

    pub fn enrich(
        &self,
        item: &ContentItem,
        config: &EnricherConfig,
    ) -> Result<EnrichmentResult, EnricherError> {
        let threshold = config.threshold.unwrap_or(0.5);
        let text = &item.content;

        let mut all_entities = Vec::new();
        all_entities.extend(find_email_entities(text));
        all_entities.extend(find_url_entities(text));
        all_entities.extend(find_date_entities(text));
        all_entities.extend(find_title_case_entities(text, threshold));

        let filtered: Vec<Entity> = all_entities.into_iter()
            .filter(|e| e.confidence >= threshold)
            .collect();
        let deduplicated = deduplicate_entities(filtered);

        let avg_conf = if deduplicated.is_empty() {
            0.0
        } else {
            deduplicated.iter().map(|e| e.confidence).sum::<f64>() / deduplicated.len() as f64
        };

        let entity_count = deduplicated.len();
        let mut type_counts = HashMap::new();
        for entity in &deduplicated {
            *type_counts.entry(format!("{:?}", entity.entity_type)).or_insert(0u64) += 1;
        }

        let mut fields = HashMap::new();
        fields.insert("entities".to_string(), serde_json::json!(deduplicated));
        fields.insert("entity_count".to_string(), serde_json::json!(entity_count));
        fields.insert("entity_type_counts".to_string(), serde_json::json!(type_counts));

        let mut metadata = HashMap::new();
        metadata.insert("provider".to_string(), serde_json::json!(PROVIDER_ID));
        metadata.insert("threshold".to_string(), serde_json::json!(threshold));
        metadata.insert("mode".to_string(), serde_json::json!("pattern_based"));

        Ok(EnrichmentResult {
            fields,
            confidence: avg_conf,
            metadata: Some(metadata),
        })
    }

    pub fn applies_to(&self, schema: &SchemaRef) -> bool {
        let text_schemas = ["text", "article", "document", "content", "post", "message", "note"];
        let name_lower = schema.name.to_lowercase();
        text_schemas.iter().any(|s| name_lower.contains(s))
    }

    pub fn cost_estimate(&self, item: &ContentItem) -> CostEstimate {
        let char_count = item.content.len();
        let duration_ms = (char_count / 1000).max(10) as u64;
        CostEstimate {
            tokens: None,
            api_calls: Some(0),
            duration_ms: Some(duration_ms),
        }
    }
}
